### PR TITLE
fix(ctb): fix abi/storage snapshot script

### PR DESCRIPTION
**Pull request by Inphi**
fixes https://github.com/ethereum-optimism/client-pod/issues/250

- Fix abi/storage snapshot checks in CI to detect ABI modifications.
- Snapshot script fixes:
  - Handle multiple contracts per source file. This introduces a couple new contracts, including "Spacer" contracts, to snapshots.
  - Detect when solc versions generates different ABIs for the same contract.
   - Remove `type` from the storageLayout schema to prevent spurious changes to ast ids
   - Fix the storageLayout `label` field to use solidity identifier names rather than its type.

## Summary by CodeRabbit

- **Refactor**
  - Improved the efficiency of snapshot difference checking in CI/CD pipeline.
  - Enhanced contract source file handling and artifact processing in snapshot generation.

- **Bug Fixes**
  - Fixed issues with ABI and storage layout snapshot generation for contracts.

- **Documentation**
  - Updated internal documentation to reflect changes in snapshot generation logic.

### CODE REVIEW

File: packages/contracts-bedrock/scripts/generate-snapshots.ts [hunk 1]
```diff
@@ -54,82 +64,85 @@ const sortKeys = (obj: any) => {
     )
 }
 
+// ContractName.0.9.8.json -> ContractName.sol
+// ContractName.json -> ContractName.sol
+const parseArtifactName = (artifactVersionFile: string): string => {
+  return artifactVersionFile.match(/(.*?)\.([0-9]+\.[0-9]+\.[0-9]+)?/)[1]
+}
```
[the remark this answers is not in the document]
Comment (tynes) [reply packages/contracts-bedrock/scripts/generate-snapshots.ts]: @Inphi I think this is a legit suggestion, what do you think?

[the remark this answers is not in the document]
Comment (refcell) [reply packages/contracts-bedrock/scripts/generate-snapshots.ts]: You're incorrect, this is a bot.

[the remark this answers is not in the document]
Comment (Inphi) [reply packages/contracts-bedrock/scripts/generate-snapshots.ts]: It's ok. Applied the bot's suggestion.

**Review by tynes — APPROVED** [review]

**Comment by tynes** [comment]
Lets merge this asap when the tests all pass and then rebase all of the open solidity PRs on top to reduce the amount of noise in each solidity PR

**Pull request merged**